src/core/mempool.rs

```diff
@@ -1832,19 +1844,59 @@ impl MemPoolDB {
             num_written += nw;
 
             if query.tx_buf_ptr >= query.tx_buf.len() {
+                if query.corked {
+                    // we're done
+                    test_debug!(
+                        "Finished streaming txs; last page was {:?}",
+                        &query.last_randomized_txid
+                    );
+                    break;
+                }
+
+                if query.num_txs >= query.max_txs {
+                    // no more space in this stream
+                    test_debug!(
+                        "No more space in this query after {:?}",
+                        &query.last_randomized_txid
+                    );
+
+                    // send the next page ID
+                    query.tx_buf_ptr = 0;
+                    query.tx_buf.clear();
+                    query.corked = true;
+
+                    test_debug!(
+                        "Cork tx stream with next page {}",
+                        &query.last_randomized_txid
+                    );
+                    query
+                        .last_randomized_txid
+                        .consensus_serialize(&mut query.tx_buf)
+                        .map_err(ChainstateError::CodecError)?;
+                    continue;
+                }
+
                 // load next
-                let (mut next_txs, next_last_randomized_txid_opt) = self
+                let (mut next_txs, next_last_randomized_txid_opt, num_rows_visited) = self
                     .find_next_missing_transactions(
                         &query.tx_query,
                         query.height,
                         &query.last_randomized_txid,
                         1,
-                        MAX_BLOOM_COUNTER_TXS.into(),
+                        query.max_txs.saturating_sub(query.num_txs),
                     )?;
+
+                test_debug!(
+                    "Visited {} rows from {}, num_txs = {}, max txs = {}",
+                    num_rows_visited,
+                    &query.last_randomized_txid,
+                    query.num_txs,
+                    query.max_txs
+                );
```

Comment:
Can this also be a `debug!` log, with something more descriptive than `Visited` as it's message, perhaps:

```suggestion
                debug!(
                    "Streaming mempool propagation stepped";
                    "rows_visited" => num_rows_visited,
                    "last_rand_txid" => %query.last_randomized_txid,
                    "num_txs" => query.num_txs,
                    "max_txs" => query.max_txs
                );
```

Reply:
Yup, changed